fms/models/granite_v4.py

```diff
@@ -0,0 +1,623 @@
+import logging
+import math
+import re
+from dataclasses import dataclass
+from typing import Any, Mapping, Optional, Tuple
+from typing_extensions import Unpack
+
+import torch
+import torch.nn as nn
+
+from fms import models
+from fms.distributed.strategy import DistributedStrategy, NoOpStrategy
+from fms.modules.attention import (
+    AttentionKwargs,
+    MultiHeadAttention,
+    get_attention_type,
+)
+from fms.modules.feedforward import GatedLinearUnit
+from fms.modules.layernorm import LayerNormParameterized
+from fms.modules.linear import get_linear_type
+from fms.modules.positions import RotaryEmbedding
+from fms.utils import serialization
+from fms.utils.activation import str_to_activation
+from fms.utils.config import ModelConfig
+from fms.utils.headless import gather_outputs
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class GraniteMoeHybridConfig(ModelConfig):
```

Comment:
If this is supposed to support MoE models (even if 8B is dense), the config lacks. We should at least add a comment to mention that this will be addressed in a future PR. 
num_experts
num_experts_per_tok
router_aux_loss_coef
Expert-specific parameters